### ISSUE TITLE
Estimating fee in TransactionBuilder when redeeming from script

### ISSUE THREAD

**Issue by dshibaev** [Emurgo/cardano-serialization-lib#375]
Both `.add_inputs_from()` and `.add_change_if_needed()` are run on TransactionBuilder before the redeemers are added to the TransactionBody. As a result they don't capture the fees to run the scripts that set on the redeemers

Alternatively setting the fee manually with `.set_fee() `on the TransactionBuilder I get this error when then running `.add_inputs_from()` and `.add_change_if_needed()`: Cannot calculate change if fee was explicitly specified

seems like its not possible to use `.add_inputs_from()` and `.add_change_if_needed()` when redeeming from a script

**Comment by vsubhuman** [comment]
Yes, the `TransactionBuilder` does not support script inputs properly at the moment, have to build or at least finish the tx manually for now. There's a plan to add this functionality with some helpers, similarly how it's been done for the minting. E.g. you'd be adding an input along with the script, datums, redeemers, and it's estimated units, etc. and then the builder provides both the fee/change calculation and the witness-set that includes all the necessary data at the end.

This is still to be done, but is planned to be added in the next week or two. Note that you must switch to use `tx_builder.build_tx()` function instead of `tx_builder.build()` - that new function is returning the full `Transaction` object with the witness set and auxiliary data included.

**Comment by aada-finance** [comment]
@vsubhuman wanted to follow up. Is there a plan to get solve this?

**Comment by lemmon-714** [comment]
@vsubhuman I second that - alternatively/additionally: Could you share some pointers what the blockers are? Might be simpler for us trying to fix the source instead of cobbling javascript around the base building blocks

**Comment by javimaravillas** [comment]
Hi all,
@vsubhuman will be working in this task and we expect to have it completed by Monday 16th if we don't face any big impediment during the road.

**Comment by AdamMachera** [comment]
Can we ask for example of how to build transaction that passes datum and redeemer so we can check if for this scenario fee is calculated properly. I would like to have a reference to know e.g. if I lock a NFT/ADA in SC, I will be able to redeem it.

**Comment by vsubhuman** [comment]
> Can we ask for example of how to build transaction that passes datum and redeemer

This is currently in progress and the new functionality in the tx builder will be available next week
